crates/astria-bridge-signer/src/verifier.rs

```diff
@@ -0,0 +1,312 @@
+        };
+        let actions = get_expected_actions_from_tx_info(
+            self.provider.clone(),
+            tx_hash,
+            event_index,
+            bridge_address,
+            act.fee_asset.clone(),
+            memo.rollup_block_number,
+            GetterConfiguration::Ics20Withdrawal(&act.denom),
+        )
+        .await
+        .wrap_err("failed to get expected actions from ics20 withdrawal tx info")?;
+        Ok(actions)
+    }
+}
+
+async fn get_expected_actions_from_tx_info(
+    provider: Arc<Provider<Http>>,
+    tx_hash: H256,
+    event_index: usize,
+    bridge_address: Address,
+    fee_asset: Denom,
+    rollup_block_number: u64,
+    getter_configuration: GetterConfiguration<'_>,
+) -> eyre::Result<Vec<Action>> {
+    let tx = provider
+        .get_transaction_receipt(tx_hash)
+        .await
+        .wrap_err("failed to get transaction")?
+        .ok_or_eyre("transaction not found")?;
+    let log = tx
+        .logs
+        .get(event_index)
+        .ok_or_eyre(format!("log at {event_index} not found"))?;
+    let contract_address = log.address;
+    let contract = astria_bridge_contracts::i_astria_withdrawer::IAstriaWithdrawer::new(
+        contract_address,
+        provider.clone(),
+    );
+
+    let mut getter_builder = astria_bridge_contracts::GetWithdrawalActionsBuilder::new()
+        .provider(provider.clone())
+        .contract_address(contract_address)
+        .bridge_address(bridge_address)
+        .fee_asset(fee_asset)
+        .use_compat_address(false);
+
+    match getter_configuration {
+        GetterConfiguration::SequencerWithdrawal => {
+            let sequencer_asset_str = contract
+                .base_chain_asset_denomination()
+                .await
+                .wrap_err("failed to get base chain asset denomination")?;
+            let sequencer_asset = Denom::from_str(&sequencer_asset_str)
+                .wrap_err("failed to parse base chain asset denomination")?;
+            getter_builder = getter_builder.set_sequencer_asset_to_withdraw(Some(sequencer_asset));
+        }
+        GetterConfiguration::Ics20Withdrawal(denom) => {
+            getter_builder =
+                getter_builder.set_ics20_asset_to_withdraw(denom.as_trace_prefixed().cloned());
+        }
+    }
+
+    let getter = getter_builder
+        .try_build()
+        .await
+        .wrap_err("failed to build getter")?;
+    let block = provider
+        .get_block(rollup_block_number)
+        .await
+        .wrap_err("failed to get block")?
+        .ok_or_eyre("block not found")?;
+    let actions: Vec<Action> = getter
+        .get_for_block_hash(block.hash.ok_or_eyre("block hash is None")?)
+        .await
+        .wrap_err("failed to get withdraw actions for block hash")?
+        .into_iter()
+        .filter_map(std::result::Result::ok)
+        .collect();
+    Ok(actions)
+}
+
+fn parse_rollup_withdrawal_event_id(event_id: &str) -> eyre::Result<(H256, usize)> {
```

Comment:
Unclear what this is supposed to be parsing. Is this returning a withdrawal event ID? This looks more like a `struct TxHashWithEventIndex { tx_hash: H256, index: usize, }`.

We should also type the errors (enum) and impl `FromStr`.